files/default/lib/packer_pipeline.rb

```diff
@@ -0,0 +1,80 @@
+#!/opt/chef/embedded/bin/ruby
+require 'json'
+require 'octokit'
+require 'faraday-http-cache'
+
+# Github API caching
+stack = Faraday::RackBuilder.new do |builder|
+  builder.use Faraday::HttpCache, serializer: Marshal, shared_cache: false
+  builder.use Octokit::Response::RaiseError
+  builder.adapter Faraday.default_adapter
+end
+Octokit.middleware = stack
+
+# Library to process github payload for Packer Template Pipeline
+class PackerPipeline
+  # find the changed files associated with the PR
+  def self.changed_files(json)
+    github = Octokit::Client.new(access_token: ENV['GITHUB_TOKEN'], auto_paginate: true)
+    repo_path = json['repository']['full_name']
+    issue_number = json['number']
+    github.pull_request_files(repo_path, issue_number)
+  end
+
+  # If file is a template, return it in an array.
+  # If it receives a script, then it locates all
+  # templates that use that script and returns them in an array.
+  def self.find_dependent_templates(file)
+    return [File.basename(file)] if file =~ /.json$/
+
+    file = File.basename file
+    dependent_templates = []
+
+    # go to dir containing all images
+    Dir.chdir(ENV['PACKER_TEMPLATES_DIR'] || './bento/packer')
+
+    # Find if a shell script is referenced
+    # iterate through images and look whether they refer to file
+    Dir.glob('*.json') do |t|
+      t_data = JSON.parse(File.read(t))
+
+      # .dig fails if it tries to index an array with a string, which is ridiculous.
+      # This will fail on other json files that start with in an array.
+      # I suppose this helps more in testing than anything, since there aren't going
+      # to be other json files in the repo.
+      next unless t_data.class.name == 'Hash'
+
+      scripts = t_data.dig('provisioners', 0, 'scripts')
+
+      # .dig returns nil if it doesn't find that path in the hash.
+      next if scripts.nil?
+
+      dependent_templates << t if scripts.any? { |f| f.include? file }
+    end
+    dependent_templates
+  end
+
+  def self.start
+    d = JSON.parse(STDIN.read)
+
+    # Iterate through all the changed files and get an array of affected templates.
+    # find_dependent_templates always returns an array of strings, so I use
+    # .reduce to stitch those arrays together.
+    templates = PackerPipeline.changed_files(d).map do |f|
+      find_dependent_templates(f.filename)
+    end.reduce(:+).uniq
```

Comment:
Explained a bit more and also added a gist link for example explaining what is happening, in the commit comment.